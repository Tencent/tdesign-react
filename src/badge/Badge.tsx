import React, { forwardRef } from 'react';
import classNames from 'classnames';
import { StyledProps } from '../common';
import useConfig from '../hooks/useConfig';
import { TdBadgeProps } from './type';
import { badgeDefaultProps } from './defaultProps';

export interface BadgeProps extends TdBadgeProps, StyledProps {}

const Badge: React.FC<BadgeProps> = forwardRef<HTMLSpanElement, BadgeProps>(
  (
    { color, dot, maxCount, count, size, shape, showZero, offset, className, content, children, style, ...restProps },
    ref,
  ) => {
    const { classPrefix } = useConfig();

    const childNode = content || children;

    const badgeClassName = classNames(
      !childNode && `${classPrefix}-badge--static`,
      dot ? `${classPrefix}-badge--dot` : `${classPrefix}-badge--${shape}`,
      size === 'small' && `${classPrefix}-size-s`,
      !childNode && className,
    );
    const getDisplayCount = () => {
      if (typeof count === 'number' && count > maxCount) {
        return `${maxCount}+`;
      }
      return count;
    };

    let isHidden = !count;
    if (typeof count === 'number') {
      isHidden = count < 1 && !showZero;
    }

    const getStyle = () => {
      const mergedStyle: React.CSSProperties = { ...style };
      if (color) mergedStyle.backgroundColor = color;
      if (offset) {
        if (offset[0]) {
          mergedStyle.right = +offset[0];
        }
        if (offset[1]) {
          mergedStyle.marginTop = +offset[1];
        }
      }
      return mergedStyle;
    };

    const badge = !isHidden ? (
      <span {...(childNode ? {} : restProps)} className={badgeClassName} style={getStyle()}>
        {!dot ? getDisplayCount() : null}
      </span>
    ) : null;

    if (!childNode) return badge;

    return (
      <span {...restProps} className={classNames(`${classPrefix}-badge`, className)} ref={ref}>
        {childNode}
        {badge}
      </span>
    );
  },
);

Badge.displayName = 'Badge';
Badge.defaultProps = badgeDefaultProps;

export default Badge;
