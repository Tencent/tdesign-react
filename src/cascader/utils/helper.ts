import { TreeNode, CascaderContextType, TreeNodeValue, TreeOptionData } from '../interface';

/**
 * 获取node的全部路径
 * @param node
 * @returns
 */
export function getFullPathLabel(node: TreeNode) {
  return node
    .getPath()
    .map((node: TreeNode) => node.label)
    .join('/');
}

/**
 * treeValue计算方法
 * @param value
 * @returns
 */
export const getTreeValue = (value: CascaderContextType['value']) => {
  let treeValue: TreeNodeValue[] = [];
  if (Array.isArray(value)) {
    if (value.length > 0 && typeof value[0] === 'object') {
      treeValue = (value as TreeOptionData[]).map((val) => val.value);
    } else if (value.length) {
      treeValue = value as TreeNodeValue[];
    }
  } else if (value) {
    if (typeof value === 'object') {
      treeValue = [(value as TreeOptionData).value];
    } else {
      treeValue = [value];
    }
  }
  return treeValue;
};

/**
 * 判断node是否在页面上
 * @param node Node
 */
export function isInPage(node: Node) {
  return node === document.body ? false : document.body.contains(node);
}

export default {
  getFullPathLabel,
};
