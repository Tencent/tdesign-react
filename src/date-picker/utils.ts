import chunk from 'lodash/chunk';
import { DateValue } from './type';
import { Styles } from '../common';

/**
 * next Month
 * @since version V1.0.0
 * @param { Date } date
 * @returns {Date}
 */
export const nextMonth = (date: Date) => {
  const isPassYear = date.getMonth() === 11;
  const newMonth = isPassYear ? 0 : date.getMonth() + 1;
  const year = isPassYear ? date.getFullYear() + 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * prev Month
 * @since version V1.0.0
 * @param {Date} date
 * @returns {Date}
 */
export const prevMonth = (date: Date) => {
  const passYear = date.getMonth() === 0;
  const newMonth = passYear ? 11 : date.getMonth() - 1;
  const year = passYear ? date.getFullYear() - 1 : date.getFullYear();
  return new Date(year, newMonth, 1);
};

/**
 * dateIndexOf
 * @since version V1.0.0
 * @param {Date} date
 * @param {Date[]} arrDates
 * @returns {Number}
 */
export function dateIndexOf(date: Date, arrDates: DateValue[]): number {
  const curDate = new Date(date);
  const curTime = curDate.setHours(0, 0, 0, 0);
  const selectedDates = arrDates.map((d: Date) => {
    const dt = new Date(d);
    const dtime = dt.setHours(0, 0, 0, 0);
    return dtime;
  });

  // console.log('curDate', selectedDates, curDate, selectedDates.indexOf(curTime));

  return selectedDates.indexOf(curTime);
}

/**
 * Empty value
 * @since version V1.0.0
 * @param mixedVar
 * @returns {Boolean}
 * @example
 * // example 1: empty(null)
 * // returns 1: true
 * // example 2: empty(undefined)
 * // returns 2: true
 * // example 3: empty([])
 * // returns 3: true
 * // example 4: empty({})
 * // returns 4: true
 * // example 5: empty({'fn' : function () {} })
 * // returns 5: false
 */
export function empty(mixedVar: any): mixedVar is undefined | null | false | '' | {} {
  let undef: undefined;
  const emptyValues = [undef, null, false, ''];
  for (let i = 0, len = emptyValues.length; i < len; i++) {
    if (mixedVar === emptyValues[i]) {
      return true;
    }
  }

  if (typeof mixedVar === 'object') {
    return Object.keys(mixedVar).length < 1;
  }

  return false;
}

/**
 * To Dashed from Camel Case
 * @since version V1.0.0
 * @param {String} strCamelCase
 * @returns {string}
 * @example
 * toDash('strCamelCase') === 'str-camel-case'
 */
export function toDash(strCamelCase: string): string {
  return String(strCamelCase)
    .split(/(?=[A-Z])/)
    .join('-')
    .toLowerCase();
}

export function setStyles(el: HTMLElement | SVGElement, styles: Styles) {
  Object.entries(styles).forEach(([prop, val = null]) => {
    if (empty(val)) {
      el.style.removeProperty(prop);
      return;
    }
    const [value, pri = ''] = String(val).split('!');

    el.style.setProperty(toDash(prop), value, pri);
  });
}

/**
 * 首字母大写
 * @param {String} str 目标字符串
 * @returns {String}
 */
export function firstUpperCase(str: string): string {
  if (!str) return str;
  return str[0].toUpperCase().concat(str.substring(1, str.length));
}

interface DateObj {
  year: number;
  month: number;
}

/**
 * 返回指定年、月的第一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getFirstDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, 1);
}

/**
 * 返回指定年、月的天数
 * @param {Object} { year, month }
 * @returns {Number}
 */
function getDaysInMonth({ year, month }: DateObj): number {
  return new Date(year, month + 1, 0).getDate();
}

/**
 * 返回指定年、月的最后一天日期
 * @param {Object} { year, month }
 * @returns {Date}
 */
function getLastDayOfMonth({ year, month }: DateObj): Date {
  return new Date(year, month, getDaysInMonth({ year, month }));
}

function isSameYear(date1: Date, date2: Date) {
  return date1.getFullYear() === date2.getFullYear();
}

function isSameMonth(date1: Date, date2: Date) {
  return isSameYear(date1, date2) && date1.getMonth() === date2.getMonth();
}

function isSameDate(date1: Date, date2: Date) {
  return isSameMonth(date1, date2) && date1.getDate() === date2.getDate();
}

/**
 * 是否是某法范围内的日期，精确到日
 * @param {Date} value 目标日期
 * @param {Object} { start, end } 范围
 * @returns {Boolean}
 */
function isBetween(
  value: { getFullYear: () => number; getMonth: () => number; getDate: () => number },
  { start, end }: { start: any; end: any },
): boolean {
  const date = new Date(value.getFullYear(), value.getMonth(), value.getDate());

  const startTime = new Date(start.getFullYear(), start.getMonth(), start.getDate());
  const endTime = new Date(end.getFullYear(), end.getMonth(), end.getDate());

  return startTime <= date && endTime >= date;
}

/**
 * 比较两个日期对象的时间戳
 * @param {Date} date1 日期1
 * @param {Date} date2 日期2
 * @returns {Number} 返回 date1.getTime() - date2.getTime() 的差值
 */
function compareAsc(date1: { getTime: () => any }, date2: Date): number {
  const d1 = date1.getTime();
  const d2 = date2.getTime();

  if (d1 < d2) return -1;
  if (d1 > d2) return 1;
  return 0;
}

/**
 * 比较两个 Date 是否是同一天 或则 同一月 或则 同一年
 * @param {Date} date1 比较的日期
 * @param {Date} date2 比较的日期
 * @param {String} type 比较类型，默认比较到『日』 date|month|year
 * @returns {Boolean}
 */
export function isSame(date1: Date, date2: Date, type = 'date'): boolean {
  const func = {
    isSameYear,
    isSameMonth,
    isSameDate,
  };
  return func[`isSame${firstUpperCase(type)}`](date1, date2);
}

export function outOfRanges(d: Date, min: any, max: any) {
  return (min && compareAsc(d, min) === -1) || (max && compareAsc(d, max) === 1);
}

/**
 * @returns {Date} 当天零点的日期对象
 */
export function getToday(): Date {
  const now = new Date();
  return new Date(now.getFullYear(), now.getMonth(), now.getDate(), 0, 0, 0);
}

/**
 * 返回日期对象的年、月、日、小时、分钟、秒、12小时制标识
 * @param {Date} date
 * @returns {Object}
 */
export function getDateObj(date: Date) {
  if (!(date instanceof Date)) {
    throw new Error('date must be Date object');
  }
  return {
    year: date.getFullYear(),
    month: date.getMonth(),
    date: date.getDate(),
    hours: date.getHours(),
    minutes: date.getMinutes(),
    seconds: date.getSeconds(),
    meridiem: date.getHours() > 11 ? 'PM' : 'AM',
  };
}

/**
 * 设置日期对象的时间部分
 * @param {Date} d 日期
 * @param {Number} hour 小时
 * @param {Number} min 分钟
 * @param {Number} sec 秒
 * @returns {Date} 一个新的date
 */
export function setDateTime(d: Date, hour: number, min: number, sec: number): Date {
  // eslint-disable-next-line
  const { year, month, date } = getDateObj(d);
  return new Date(year, month, date, hour, min, sec, 0);
}

/**
 * 增加月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function subtractMonth(date: Date, num: any): Date {
  const day = date.getDate();
  const newDate = new Date(date);

  let monthNum = num;
  // eslint-disable-next-line no-plusplus
  while (monthNum--) {
    newDate.setDate(0);
  }
  newDate.setDate(day);
  return newDate;
}

/**
 * 减月份
 * @param {Date} date 起始日期
 * @param {Number} num 月份数
 * @returns {Date}
 */
export function addMonth(date: Date, num: number): Date {
  let monthNum = num;
  if (num < 0) monthNum = 0;
  const newDate = new Date(date);
  newDate.setMonth(date.getMonth() + monthNum);
  return newDate;
}

export interface OptionsType {
  firstDayOfWeek: number;
  disableDate: Function;
  minDate: Date;
  maxDate: Date;
  monthText?: string;
}

export function getWeeks(
  { year, month }: { year: number; month: number },
  { firstDayOfWeek, disableDate = () => false, minDate, maxDate }: OptionsType,
) {
  const prependDay = getFirstDayOfMonth({ year, month });

  const appendDay = getLastDayOfMonth({ year, month });

  const maxDays = getDaysInMonth({ year, month });

  const daysArr = [];

  const today = getToday();

  for (let i = 1; i <= maxDays; i++) {
    const currentDay = new Date(year, month, i);

    daysArr.push({
      text: i,
      active: false,
      value: currentDay,
      disabled: disableDate(currentDay) || outOfRanges(currentDay, minDate, maxDate),
      now: isSame(today, currentDay),
      firstDayOfMonth: i === 1,
      lastDayOfMonth: i === maxDays,
      type: 'current-month',
    });
  }

  if (prependDay.getDay() !== firstDayOfWeek - 1) {
    prependDay.setDate(0); // 上一月

    const FLAG = true;
    while (FLAG) {
      daysArr.unshift({
        text: prependDay.getDate().toString(),
        active: false,
        value: new Date(prependDay),
        disabled: disableDate(prependDay) || outOfRanges(prependDay, minDate, maxDate),
        additional: true, // 非当前月
        type: 'prev-month',
      });
      prependDay.setDate(prependDay.getDate() - 1);
      if (prependDay.getDay() === Math.abs(firstDayOfWeek + 6) % 7) break;
    }
  }

  const LEN = 42; // 显示6周

  while (daysArr.length < LEN) {
    appendDay.setDate(appendDay.getDate() + 1);
    daysArr.push({
      text: appendDay.getDate(),
      active: false,
      value: new Date(appendDay),
      disabled: disableDate(appendDay) || outOfRanges(appendDay, minDate, maxDate),
      additional: true, // 非当前月
      type: 'next-month',
    });
  }

  return chunk(daysArr, 7);
}

export function getYears(year: number, { disableDate = () => false, minDate, maxDate }: OptionsType) {
  const startYear = parseInt((year / 10).toString(), 10) * 10;
  const endYear = startYear + 9;

  const yearArr = [];

  const today = getToday();

  for (let i = startYear; i <= endYear; i++) {
    const date = new Date(i, 1);

    let disabledMonth = 0;
    let outOfRangeMonth = 0;

    for (let j = 0; j < 12; j++) {
      const d = new Date(i, j);
      if (disableDate(d)) disabledMonth += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeMonth += 1;
    }

    yearArr.push({
      value: date,
      now: isSame(date, today, 'year'),
      disabled: disabledMonth === 12 || outOfRangeMonth === 12,
      active: false,
      text: `${date.getFullYear()}`,
    });
  }

  return chunk(yearArr, 4);
}

export function getMonths(year: number, { disableDate = () => false, minDate, maxDate, monthText = '' }: OptionsType) {
  const monthArr = [];

  const today = getToday();

  for (let i = 0; i <= 11; i++) {
    const date = new Date(year, i);

    let disabledDay = 0;
    let outOfRangeDay = 0;

    const daysInMonth = getDaysInMonth({ year, month: i });

    for (let j = 1; j <= daysInMonth; j++) {
      const d = new Date(year, i, j);
      if (disableDate(d)) disabledDay += 1;
      if (outOfRanges(d, minDate, maxDate)) outOfRangeDay += 1;
    }

    monthArr.push({
      value: date,
      now: isSame(date, today, 'month'),
      disabled: disabledDay === daysInMonth || outOfRangeDay === daysInMonth,
      active: false,
      text: `${date.getMonth() + 1} ${monthText}`,
    });
  }

  return chunk(monthArr, 4);
}

interface DateTime {
  active: boolean;
  highlight: boolean;
  startOfRange: boolean;
  endOfRange: boolean;
  value: Date;
}

export function flagActive(data: any[], { ...args }: any) {
  const { start, end, type = 'date' } = args;

  if (!end) {
    return data.map((row: any[]) =>
      row.map((item: DateTime) => {
        Object.assign(item, { active: isSame(item.value, start, type) });
        return item;
      }),
    );
  }

  return data.map((row: any[]) =>
    row.map((item: DateTime) => {
      const date = item.value;
      const isStart = isSame(start, date, type);
      const isEnd = isSame(end, date, type);
      Object.assign(item, {
        active: isStart || isEnd,
        highlight: isBetween(date, { start, end }),
        startOfRange: isStart,
        endOfRange: isEnd,
      });
      return item;
    }),
  );
}
