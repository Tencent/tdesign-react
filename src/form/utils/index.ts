import type { NamePath } from '../type';

// 获取 formMap 管理的数据
export function getMapValue(name: NamePath, formMapRef: React.MutableRefObject<Map<any, any>>) {
  if (!formMapRef.current) return;

  // 提取所有 map key
  const mapKeys = [...formMapRef.current.keys()];
  // 转译为字符串后比对 key 兼容数组格式
  const key = mapKeys.find((key) => String(key) === String(name));
  // 拿到 key 引用地址获取 value
  return formMapRef.current.get(key);
}

// 将数据整理成对象，数组 name 转化嵌套对象: ['user', 'name'] => { user: { name: '' } }
export function calcFieldValue(name: NamePath, value: any) {
  if (Array.isArray(name)) {
    const fieldValue = name.reduceRight((prev, curr) => {
      const arr = [];
      if (/^\d+$/.test(String(curr))) arr[curr] = prev;
      return arr.length ? arr : { [curr]: prev };
    }, value);
    return { ...fieldValue };
  }

  return { [name]: value };
}

// 通过对象数据类型获取 map 引用: { user: { name: '' } } => formMap.get(['user', 'name'])
export function travelMapFromObject(
  obj: Record<any, any>,
  formMapRef: React.MutableRefObject<Map<any, any>>,
  callback: Function,
) {
  for (const [mapName, formItemRef] of formMapRef.current.entries()) {
    // 支持嵌套数据结构
    if (Array.isArray(mapName)) {
      // 创建唯一临时变量 symbol
      const symbol = Symbol('name');
      let fieldValue = null;

      for (let i = 0; i < mapName.length; i++) {
        const item = mapName[i];
        if (Reflect.has(fieldValue || obj, item)) {
          fieldValue = Reflect.get(fieldValue || obj, item);
        } else {
          // 当反射无法获取到值则重置为 symbol
          fieldValue = symbol;
          break;
        }
      }

      // 非 symbol 说明获取到了值
      if (fieldValue !== symbol) {
        callback(formItemRef, fieldValue);
      }
    } else if (Reflect.has(obj, mapName)) {
      callback(formItemRef, obj[mapName]);
    }
  }
}
