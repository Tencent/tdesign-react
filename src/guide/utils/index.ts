import { scrollToParentVisibleArea } from './getScrollParent';
import getRelativePosition from './getRelativePosition';
import getTargetElm from './getTargetElm';
import scrollToElm from './scrollToElm';
import useMount from './useMount';

export { scrollToParentVisibleArea, getRelativePosition, getTargetElm, scrollToElm, useMount };
