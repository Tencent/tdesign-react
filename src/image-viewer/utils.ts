export const downloadFile = function (imgSrc) {
  const image = new Image();
  const name = imgSrc?.split?.('/').pop() || Math.random().toString(32).slice(2);

  image.setAttribute('crossOrigin', 'anonymous');

  image.onload = () => {
    const canvas = document.createElement('canvas');
    canvas.width = image.width;
    canvas.height = image.height;

    const context = canvas.getContext('2d');
    context.drawImage(image, 0, 0, image.width, image.height);
    canvas.toBlob((blob) => {
      const url = URL.createObjectURL(blob);
      const a = document.createElement('a');
      a.download = name;
      a.href = url;
      a.click();
      a.remove();
      URL.revokeObjectURL(url);
    });
  };
  image.src = imgSrc;
};
