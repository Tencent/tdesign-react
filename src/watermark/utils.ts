const toLowercaseSeparator = (key: string) => key.replace(/([A-Z])/g, '-$1').toLowerCase();

// style对象转字符串
export const getStyleStr = (style: React.CSSProperties): string =>
  Object.keys(style)
    .map((key: keyof React.CSSProperties) => `${toLowercaseSeparator(key)}: ${style[key]};`)
    .join(' ');
