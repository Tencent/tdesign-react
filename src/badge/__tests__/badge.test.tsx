import React from 'react';
import { testExamples, render } from '@test/utils';
import Badge from '../Badge';

// 测试组件代码 Example 快照
testExamples(__dirname);

describe('Badge 组件测试', () => {
  function renderBadge(badge) {
    const { container } = render(badge);
    return container.firstChild;
  }

  test('默认 DOM 结构', async () => {
    const badge = renderBadge(<Badge count={1}>test</Badge>);
    expect(badge.firstChild.textContent).toBe('test');
    expect(badge.lastChild.textContent).toBe('1');
    expect(badge.lastChild).toHaveClass('t-badge--circle');
    expect(badge.childNodes.length).toBe(2);
  });

  test('class', async () => {
    expect(renderBadge(<Badge count={1}>test</Badge>).lastChild).toHaveClass('t-badge--circle');
    expect(
      renderBadge(
        <Badge shape="round" count={1}>
          test
        </Badge>,
      ).lastChild,
    ).toHaveClass('t-badge--round');
    expect(
      renderBadge(
        <Badge dot count={1}>
          test
        </Badge>,
      ).lastChild,
    ).toHaveClass('t-badge--dot');
    expect(
      renderBadge(
        <Badge size="small" count={1}>
          test
        </Badge>,
      ).lastChild,
    ).toHaveClass('t-badge--circle', 't-size-s');

    // not a wrapper
    expect(renderBadge(<Badge count={1} />)).toHaveClass('t-badge--static', 't-badge--circle');
  });

  test('count', async () => {
    // count 0
    expect(renderBadge(<Badge />)).toBeNull();
    expect(renderBadge(<Badge showZero count={0} />)).toBeTruthy();

    // maxCount
    expect(renderBadge(<Badge count={100} />)).toHaveTextContent('99+');
    expect(renderBadge(<Badge count={100} maxCount={100} />)).toHaveTextContent('100');

    // count
    expect(renderBadge(<Badge count={2} />)).toHaveTextContent('2');
  });

  test('offset', async () => {
    const badge = renderBadge(
      <Badge offset={[1, 2]} count={1}>
        test
      </Badge>,
    );
    expect(badge.lastChild).toHaveStyle({ right: '1px', 'margin-top': '2px' });
  });

  test('color', async () => {
    expect(renderBadge(<Badge color="red" count={1} />)).toHaveStyle({ 'background-color': 'red' });
  });
});
