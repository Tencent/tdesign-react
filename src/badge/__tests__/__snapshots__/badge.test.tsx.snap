// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`base.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base t-size-l"
      type="button"
    >
      <span
        class="t-button__text"
      >
         
      </span>
    </button>
    <span
      class="t-badge--dot"
    />
  </span>
  <span
    class="t-badge"
  >
    解锁新徽章
    <span
      class="t-badge--dot"
    />
  </span>
  <span
    class="t-badge"
  >
    <svg
      class="t-icon t-icon-user"
      fill="none"
      height="1em"
      style="font-size: 24px;"
      viewBox="0 0 16 16"
      width="1em"
    >
      <path
        d="M11.5 5a3.5 3.5 0 11-7 0 3.5 3.5 0 017 0zm-1 0a2.5 2.5 0 10-5 0 2.5 2.5 0 005 0zM13.96 10.85c.34.16.54.5.54.87V14a.5.5 0 01-.5.5H2a.5.5 0 01-.5-.5v-2.28c0-.37.2-.7.54-.87a13.79 13.79 0 0111.92 0zM8 10.5c-1.97 0-3.83.45-5.5 1.24v1.76h11v-1.76A12.78 12.78 0 008 10.5z"
        fill="currentColor"
        fill-opacity="0.9"
      />
    </svg>
    <span
      class="t-badge--dot"
    />
  </span>
</DocumentFragment>
`;

exports[`color.jsx 1`] = `
<DocumentFragment>
  <label
    class="t-checkbox"
  >
    <input
      class="t-checkbox__former"
      type="checkbox"
      value="1"
    />
    <span
      class="t-checkbox__input"
    />
    <span
      class="t-checkbox__label"
    >
      红点样式
    </span>
  </label>
  <div>
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(244, 67, 54);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(233, 30, 99);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(156, 39, 176);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(63, 81, 181);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(33, 150, 243);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(3, 169, 244);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(0, 188, 212);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(0, 150, 136);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(76, 175, 80);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(139, 195, 74);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(205, 220, 57);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(255, 235, 59);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(255, 193, 7);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(255, 152, 0);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(255, 87, 34);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(121, 85, 72);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(158, 158, 158);"
    />
    <span
      class="t-badge--static t-badge--dot"
      style="background-color: rgb(96, 125, 139);"
    />
  </div>
</DocumentFragment>
`;

exports[`custom.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-badge"
  >
    <div
      style="width: 40px; height: 40px; background: rgb(238, 238, 238); border: 1px solid #dcdcdc; box-sizing: border-box; border-radius: 3px;"
    />
    <span
      class="t-badge--dot"
    />
  </span>
  <span
    class="t-badge"
  >
    <div
      style="width: 40px; height: 40px; background: rgb(238, 238, 238); border: 1px solid #dcdcdc; box-sizing: border-box; border-radius: 3px;"
    />
    <span
      class="t-badge--circle"
    >
      hot
    </span>
  </span>
  <span
    class="t-badge"
  >
    <div
      style="width: 40px; height: 40px; background: rgb(238, 238, 238); border: 1px solid #dcdcdc; box-sizing: border-box; border-radius: 3px;"
    />
    <span
      class="t-badge--circle"
      style="background-color: rgb(0, 168, 112);"
    >
      new
    </span>
  </span>
  <span
    class="t-badge"
  >
    <div
      style="width: 40px; height: 40px; background: rgb(238, 238, 238); border: 1px solid #dcdcdc; box-sizing: border-box; border-radius: 3px;"
    />
    <span
      class="t-badge--round"
      style="background-color: rgb(0, 82, 217);"
    >
      100
    </span>
  </span>
</DocumentFragment>
`;

exports[`number.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base t-size-l"
      type="button"
    >
      <span
        class="t-button__text"
      >
         
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base t-size-l"
      type="button"
    >
      <span
        class="t-button__text"
      >
         
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      99
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base t-size-l"
      type="button"
    >
      <span
        class="t-button__text"
      >
         
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      99+
    </span>
  </span>
</DocumentFragment>
`;

exports[`offset.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        默认
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        [10,10]
      </span>
    </button>
    <span
      class="t-badge--circle"
      style="right: 10px; margin-top: 10px;"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        [-10,10]
      </span>
    </button>
    <span
      class="t-badge--circle"
      style="right: -10px; margin-top: 10px;"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        [-10,-10]
      </span>
    </button>
    <span
      class="t-badge--circle"
      style="right: -10px; margin-top: -10px;"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        [10,-10]
      </span>
    </button>
    <span
      class="t-badge--circle"
      style="right: 10px; margin-top: -10px;"
    >
      2
    </span>
  </span>
</DocumentFragment>
`;

exports[`shape.jsx 1`] = `
<DocumentFragment>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
         circle
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        round
      </span>
    </button>
    <span
      class="t-badge--round"
    >
      99
    </span>
  </span>
</DocumentFragment>
`;

exports[`size.jsx 1`] = `
<DocumentFragment>
  <h3
    style="margin-bottom: 16px;"
  >
    1.默认大小
  </h3>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        按钮
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        按钮
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      99
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        按钮
      </span>
    </button>
    <span
      class="t-badge--circle"
    >
      99+
    </span>
  </span>
  <h3
    style="margin-bottom: 16px; margin-top: 32px;"
  >
    2.小
  </h3>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        按钮
      </span>
    </button>
    <span
      class="t-badge--circle t-size-s"
    >
      2
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        按钮
      </span>
    </button>
    <span
      class="t-badge--circle t-size-s"
    >
      99
    </span>
  </span>
  <span
    class="t-badge"
  >
    <button
      class="t-button t-button--theme-primary t-button--variant-base"
      type="button"
    >
      <span
        class="t-button__text"
      >
        按钮
      </span>
    </button>
    <span
      class="t-badge--circle t-size-s"
    >
      99+
    </span>
  </span>
</DocumentFragment>
`;
